Neos.RedirectHandler.NeosAdapter/Classes/Service/NodeRedirectionService.php

```diff
@@ -0,0 +1,218 @@
+        }
+
+        $hosts = $this->getHostPatterns($node->getContext());
+
+        // The page has been removed
+        if ($node->isRemoved()) {
+            $this->flushRoutingCacheForNode($targetNode);
+            $statusCode = (integer)$this->defaultStatusCode['gone'];
+            $this->redirectionStorage->addRedirection($targetNodeUriPath, '', $statusCode, $hosts);
+            return;
+        }
+
+        // compare the "old" node URI to the new one
+        $nodeUriPath = $this->buildUriPathForNodeContextPath($node->getContextPath());
+        // use the same regexp than the ContentContextBar Ember View
+        $nodeUriPath = preg_replace('/@[A-Za-z0-9;&,\-_=]+/', '', $nodeUriPath);
+        if ($nodeUriPath === null || $nodeUriPath === $targetNodeUriPath) {
+            // The page node path has not been changed
+            return;
+        }
+
+
+        $this->flushRoutingCacheForNode($targetNode);
+        $statusCode = (integer)$this->defaultStatusCode['redirect'];
+        $this->redirectionStorage->addRedirection($targetNodeUriPath, $nodeUriPath, $statusCode, $hosts);
+        /** @var ContentContext $contentContext */
+
+
+        $q = new FlowQuery([$node]);
+        foreach ($q->children('[instanceof TYPO3.Neos:Document]') as $childrenNode) {
+            $this->createRedirectionsForPublishedNode($childrenNode, $targetWorkspace);
+        }
+    }
+
+    /**
+     * @param ContentContext $contentContext
+     * @return array
+     */
+    protected function getHostPatterns(ContentContext $contentContext)
+    {
+        $site = $contentContext->getCurrentSite();
+        $domains = [];
+        if ($site !== null) {
+            foreach ($site->getDomains() as $domain) {
+                /** @var Domain $domain */
+                $domains[] = $domain->getHostPattern();
+            }
+        }
+        return $domains;
+    }
+
+    /**
+     * Removes all routing cache entries for the given $nodeData
+     *
+     * @param NodeInterface $node
+     * @return void
+     */
+    protected function flushRoutingCacheForNode(NodeInterface $node)
+    {
+        $nodeData = $node->getNodeData();
+        $nodeDataIdentifier = $this->persistenceManager->getIdentifierByObject($nodeData);
+        if ($nodeDataIdentifier === null) {
+            return;
+        }
+        $this->routerCachingService->flushCachesByTag($nodeDataIdentifier);
+    }
+
+    /**
+     * Creates a (relative) URI for the given $nodeContextPath removing the "@workspace-name" from the result
+     *
+     * @param string $nodeContextPath
+     * @return string the resulting (relative) URI or NULL if no route could be resolved
+     */
+    protected function buildUriPathForNodeContextPath($nodeContextPath)
+    {
+        try {
+            return $this->getUriBuilder()
+                ->uriFor('show', ['node' => $nodeContextPath], 'Frontend\\Node', 'TYPO3.Neos');
```

Comment:
since the linking service is bypassed, shortcuts are not resolved.. does that have unexpected consequences?